doc: Add a guide page about AMD perf issue (#384)
xcp-ng/xcp-ng-org: docs/guides/amd-performance-improvements.md

# Network traffic performance on AMD processors

> This page details performance improvements on AMD processors and how to deploy it on your pool.

Reviewer: What does 'it' refer to, here? It's not clear to me.

## What's the "AMD EPYC performance issue" story?

> Formally known as "AMD EPYC performance issue", it's necessary to clarify that all AMD processors are affected, not only EPYC product line. It was only identified on EPYC processors first in 2024.

Reviewer 2: Already solved :+1:

> The performance issue on AMD processors was characterized by low bandwidth on the internal VM to VM (inside the same host) and VM to host network traffic, especially when compared with Intel processors.

Reviewer: ```suggestion
The performance issue on AMD processors was characterized by low bandwidth on internal VM-to-VM traffic (inside the same host) and VM-to-host-network traffic, especially when compared with Intel processors.
```

> The network traffic is performant enough for many workloads, but not for the most demanding ones.

Reviewer: ```suggestion
The network traffic is sufficient for many workloads, but not for the most demanding ones.
```

> As often with performance problems, this is a complicated issue, likely not bound to one single cause but at least a bottleneck was removed.

Reviewer: ```suggestion
As with most performance issues, this case is complicated. The issue can stem from multiple causes, but at least one bottleneck has been eliminated.
```

## A first significant fix: uncached grant tables

### Technical description of the issue and its fix

> Due to differences between processors architectures and Xen's isolation, the Grant-Tables in Xen were not cached the same way on Intel and AMD. This patch fixes the location of the Grant-Tables mapping to change the caching from `Uncached` to `Writeback`.

Reviewer: Spelling of Grant tables seems inconsistent. We have:

- grant tables
- Grant-Tables

If possible, we shoulds stick to the same spelling across the whole document.

Author: The only spelling is 'grant-tables' now. Ok for you?

> Finally, with the help of the XenServer team, we understood that there's a specific data structure that isn't cached, and causes a significant performance hit.

Reviewer: ```suggestion
Finally, with help of the XenServer team, we realized that a specific data structure isn't cached, which causes a significant performance hit.
```

> However, this fix is _not_ useful on Intel processors but the fix can be enable as well, as there wasn't any regression noted brought by the changes.

Reviewer: ```suggestion
However, this fix is _not_ useful on Intel processors. You can enable it on Intel processors anyway, since the changes are not known to cause any regression.
```

### How to enable the fix

:::warning
The fix is only available on XCP-ng 8.3 with `xapi` >= 25.6.0-1.11.xcpng8.3 and `xen` >= 4.17.5-15.2.xcpng8.3
:::

> Xenopsd supports a new option allowing to activate the remapping of grant-tables as Writeback (`xen-platform-pci-bar-uc`).

Reviewer: ```suggestion
Xenopsd supports a new option, allowing the grant table to not be mapped as `writeback` instead of `uncacheable`.
```

> Xenopsd supports a new option allowing to activate the remapping of grant-tables as Writeback (`xen-platform-pci-bar-uc`).

Author: I fixed this line with a slightly different version. Ok for you?

> Guests need their kernel to support the feature which enables this fix (See partial list below). Windows and BSD guests were not affected by the performance problem this change solves.

Reviewer: ```suggestion
Guests need their kernel to support the feature that enables the fix (See partial list below). Windows and BSD guests were not affected by the performance problem solved by this change.
```

This change has to be opt-in, so that users be conscious of the change and also know how to revert it if any side effects remain in edge cases.

* To enable the fix pool-wide, create a file named `/etc/xenopsd.conf.d/amd-gnttab-wb.conf` with the following line:

```
xen-platform-pci-bar-uc=false
```

* Then restart the toolstack on the host: `xe-toolstack-restart`
* Then add the configuration and restart the toolstack on every other hosts of the pool

> * Then stop and start VMs so that the setting is applied to them at boot.

Author: A simple restart is not enough to apply the configuration but I modify the line to be more clear. Good for you?

> After a few months (in a future update of XCP-ng 8.3), this will become the default.

Reviewer: ```suggestion
This configuration will become default in a few months, with a future update to XCP-ng 8.3.
```

This is not the end of the way towards better performance on AMD EPYC servers, but it's significant progress!

### OS support

> Linux distributions that have recent enough kernels or apply fixes from the mainline LTS kernels are OK. Older ones are not. Some currently supported LTS distros don't have the patch yet: RHEL 8 and 9 and their derivatives are not ready yet - no effect on older distros such as Ubuntu 20.04. 

Reviewer: ```suggestion
**Ready:**
Linux distributions with recent enough kernels or distros that apply fixes from the mainline LTS kernels are OK. 

**Not ready:**

Older distros. 

Some currently-supported LTS distros don't have the patch yet. Specifically: RHEL 8 and 9 and their derivatives.

**No effect**

Older distros such as Ubuntu 20.04 are not affected.
```

The performance fix support is related to the guest kernel version. All kernel >= 5.19 work with it. LTS kernels include 5.15.189+.

* Debian:
  * Debian 11 "Bullseye" (5.10 -> _Waiting for LTS kernel to be patch_)
  * Debian 12 "Bookworm" (6.1 -> **OK**)
* Ubuntu:
  * 20.04 LTS "Focal Fossa" (5.4: _EOL_)
  * 22.04 LTS "Jammy Jellyfish" (5.15: _SOON_, HWE 6.8: **OK**)
  * 24.04 LTS "Noble Numbat" (6.8 & HWE 6.14: **OK**)
* openSUSE Leap:
  * 15.5 (5.14: _EOL_)
  * 15.6 (6.4: **OK**)
* SUSE Enterprise (LTSS):
  * SLE15 SP3 - LTSS (5.3: _Not upstream_)
  * SLE15 SP4/5 - LTSS (5.14: _Not upstream_)
  * SLE15 SP6+ (**OK**)
* RHEL (+derivates):
  * 8 (4.19: _EOL_)
  * 9 (5.14: _Not upstream_)
  * 10 (6.12: **OK**)
* Fedora: All supported: **OK** (37+)
* Alpine Linux: All supported: **OK** (v3.18+)

* _EOL_ = distro is End-of-Life
* _Not upstream_ = not covered by Linux stable project
* _SOON_: Distro needs to update its kernel

> ## A second fix: spurious interrupts

Reviewer: ```suggestion
## Fix 2: spurious interrupts
```

> ### Technical description of the issue and it fix

Reviewer: ```suggestion
### Technical description of the issue and its fix
```

Spurious interrupts are interrupts that are rised for nothing, there is no work to do.
To avoid deny of service, xen devices front-ends implement a security to avoid spurious interrupts from the backends by putting interrupt handler on hold for a short period of time, hoping that this delay will slow down the backend in its spurious interruptions frenzy.

> Sadly the spurious interrupts weren't caused by the backend but by another part of the code that consume packets outside of the interrput handler.

Reviewer: ```suggestion
Sadly, the spurious interrupts weren't caused by the backend. They were caused but by another part of the code, which consumes packets outside of the interrupt handler.
```

The real cause was fixed upstream. Don't hesitate to read the full [commit message](https://lore.kernel.org/all/20250721093316.23560-1-anthoine.bourgeois@vates.tech/) for more technical details.

### How to enable the fix

This issue was only on the front-end side meaning that only an upgrade of the guest kernel can improve the network bandwidth by lower the spurious interrupts.

> This fix can be combine with the first fix on grant-table, both are independent.

Reviewer: ```suggestion
Fix 1 (on grant-tables) and fix 2 work independently from each other. You can combine both.
```

> It doesn't warranty better performance for every processors models (See Future improvements below) but no regression with it installed was noted. 

Reviewer: ```suggestion
The fixes don't guarantee better performance for every processor model (See Future improvements) but we didn't detect any regression when applying them.
```

Also, consider making "Future improvements" a clickable link that brings the user to that section

> This fix is not specific to AMD, it can improve internal network performance on Intel processors too.

Reviewer: ```suggestion
The fix is not specific to AMD —it can improve internal network performance on Intel processors too.
```

### OS support

As the fix is very young no distribution embed this fix yet.

But the fix is upstream in the Linux kernel (v6.17) and LTS kernel are up-to-date.
Here the list of LTS kernels containing the improvement:
* v5.10.241
* v5.15.190
* v6.1.149
* v6.6.103
* v6.12.43
* v6.15.11
* v6.16.2

## Future improvements and leads

> The last improvement is not effective on every platforms, so it seems the story is not over, there is at least a bottleneck left.

Reviewer: ```suggestion
The last improvement does not work on every platform, so the story may not be over yet. At least one bottleneck remains.
```

> ## A second fix: spurious interrupts
> 
> ### Technical description of the issue and it fix
> 
> Spurious interrupts are interrupts that are rised for nothing, there is no work to do.
> To avoid deny of service, xen devices front-ends implement a security to avoid spurious interrupts from the backends by putting interrupt handler on hold for a short period of time, hoping that this delay will slow down the backend in its spurious interruptions frenzy.
> 
> Sadly the spurious interrupts weren't caused by the backend but by another part of the code that consume packets outside of the interrput handler.
> 
> The real cause was fixed upstream. Don't hesitate to read the full [commit message](https://lore.kernel.org/all/20250721093316.23560-1-anthoine.bourgeois@vates.tech/) for more technical details.
> 
> ### How to enable the fix
> 
> This issue was only on the front-end side meaning that only an upgrade of the guest kernel can improve the network bandwidth by lower the spurious interrupts.
> 
> This fix can be combine with the first fix on grant-table, both are independent.
> It doesn't warranty better performance for every processors models (See Future improvements below) but no regression with it installed was noted. 
> This fix is not specific to AMD, it can improve internal network performance on Intel processors too.
> 
> ### OS support
> 
> As the fix is very young no distribution embed this fix yet.
> 
> But the fix is upstream in the Linux kernel (v6.17) and LTS kernel are up-to-date.
> Here the list of LTS kernels containing the improvement:
> * v5.10.241
> * v5.15.190
> * v6.1.149
> * v6.6.103
> * v6.12.43
> * v6.15.11
> * v6.16.2
> 
> ## Future improvements and leads
> 
> The last improvement is not effective on every platforms, so it seems the story is not over, there is at least a bottleneck left.
> 
> The investigation continues.

Reviewer: Should we keep that part, I suggest the following:

```suggestion
The investigation goes on.
```

> The investigation continues.

Author: I'm not sure you talk about the spurious interrupts part, the future improvements part or both. As there is no option or switch to trigger on the spurious interrupts part, I agree it is more informational. The future improvements is quite empty, so not useful except as a conclusion (but again a conclusion is more suited for a blog post).
I can remove those and make it a blog post as you like. @stormi what do you think?

> The investigation continues.

Author: I'm applying Thomas's wording :+1:

* Ubuntu:
  * 20.04 LTS "Focal Fossa" (5.4: _EOL_)
  * 22.04 LTS "Jammy Jellyfish" (5.15: _SOON_, HWE 6.8: **OK**)
  * 24.04 LTS "Noble Numbat" (6.8 & HWE 6.14: **OK**)
* openSUSE Leap:
  * 15.5 (5.14: _EOL_)
  * 15.6 (6.4: **OK**)
* SUSE Enterprise (LTSS):
  * SLE15 SP3 - LTSS (5.3: _Not upstream_)
  * SLE15 SP4/5 - LTSS (5.14: _Not upstream_)
  * SLE15 SP6+ (**OK**)
* RHEL (+derivates):
  * 8 (4.19: _EOL_)
  * 9 (5.14: _Not upstream_)
  * 10 (6.12: **OK**)
* Fedora: All supported: **OK** (37+)
* Alpine Linux: All supported: **OK** (v3.18+)

* _EOL_ = distro is End-of-Life
* _Not upstream_ = not covered by Linux stable project
* _SOON_: Distro needs to update its kernel

## Fix 2: spurious interrupts

### Technical description of the issue and its fix

Spurious interrupts are interrupts that arise for no reason —there is no work to do.
To avoid denial-of-service, xen devices front-ends implement security measures to avoid spurious interrupts from the backends. To do that, front-ends put interrupt handlers on hold for a short period of time. With this delay, front-ends can hopefully mitigate the flood of spurious interruptions caused by the backend.

> Sadly, the spurious interrupts weren't caused by the backend. They were caused but by another part of the code, which consumes packets outside of the interrupt handler.

Reviewer 2: ```suggestion
Sadly, the spurious interrupts weren't caused by the backend. They were caused by another part of the code, which consumes packets outside of the interrupt handler.
```

The real cause was fixed upstream. Don't hesitate to read the full [commit message](https://lore.kernel.org/all/20250721093316.23560-1-anthoine.bourgeois@vates.tech/) for more technical details.

### How to enable the fix

> This issue occurred only on the front-end side. This means that the only way to improve network bandwith is to upgrade the guest kernel, which limits spurious interrupts.

Reviewer 2: ```suggestion
This issue occurred only on the front-end side. This means that the only way to improve network bandwidth is to upgrade the guest kernel, which limits spurious interrupts.
```

xen-platform-pci-bar-uc=false
```

1. Restart the toolstack on the host: `xe-toolstack-restart`
2. Add the configuration and restart the toolstack on every other host of the pool

> 3. **Stop and start VMs** to apply the configuration during the new boot (a simple reboot in not enough).

Reviewer 2: s/in not/is not/

Author: Same, good catch.
